File: ios/Classes/nativeModel/EnrichedNativeError.swift
import Foundation

/// Enriched error information from native operations
/// Preserves structured error data for proper error handling in Dart
@objc public class EnrichedNativeError: NSObject {
    @objc public let message: String
    @objc public let code: String
    @objc public let nativeError: String?
    @objc public let nativeErrorCode: Int?
    @objc public let nativeErrorDomain: String?
    @objc public let context: [String: Any]?
    @objc public let timestamp: String
    @objc public let nativeStackTrace: String?
    @objc public let operationId: String
    @objc public let instanceId: String
    @objc public let queue: String
    
    @objc public init(
        message: String,
        code: String,
        nativeError: String? = nil,
        nativeErrorCode: Int? = nil,
        nativeErrorDomain: String? = nil,
        context: [String: Any]? = nil,
        timestamp: String,
        nativeStackTrace: String? = nil,
        operationId: String,
        instanceId: String,
        queue: String
    ) {
        self.message = message
        self.code = code
        self.nativeError = nativeError
        self.nativeErrorCode = nativeErrorCode
        self.nativeErrorDomain = nativeErrorDomain
        self.context = context
        self.timestamp = timestamp
        self.nativeStackTrace = nativeStackTrace
        self.operationId = operationId
        self.instanceId = instanceId
        self.queue = queue
        super.init()
    }
    
    /// Convert to dictionary for channel communication
    @objc public func toDictionary() -> [String: Any] {
        var dict: [String: Any] = [
            "message": message,
            "code": code,
            "timestamp": timestamp,
            "operationId": operationId,
            "instanceId": instanceId,
            "queue": queue,
        ]
        
        if let nativeError = nativeError {
            dict["nativeError"] = nativeError
        }
        
        if let nativeErrorCode = nativeErrorCode {
            dict["nativeErrorCode"] = nativeErrorCode
        }
        
        if let nativeErrorDomain = nativeErrorDomain {
            dict["nativeErrorDomain"] = nativeErrorDomain
        }
        
        if let context = context {
            dict["context"] = context
        }
        
        if let nativeStackTrace = nativeStackTrace {
            dict["nativeStackTrace"] = nativeStackTrace
        }
        
        return dict
    }
    
    /// Create from dictionary received from channel
    @objc public static func fromDictionary(_ dict: [String: Any]) -> EnrichedNativeError? {
        guard let message = dict["message"] as? String,
              let code = dict["code"] as? String,
              let timestamp = dict["timestamp"] as? String,
              let operationId = dict["operationId"] as? String,
              let instanceId = dict["instanceId"] as? String,
              let queue = dict["queue"] as? String else {
            return nil
        }
        
        return EnrichedNativeError(
            message: message,
            code: code,
            nativeError: dict["nativeError"] as? String,
            nativeErrorCode: dict["nativeErrorCode"] as? Int,
            nativeErrorDomain: dict["nativeErrorDomain"] as? String,
            context: dict["context"] as? [String: Any],
            timestamp: timestamp,
            nativeStackTrace: dict["nativeStackTrace"] as? String,
            operationId: operationId,
            instanceId: instanceId,
            queue: queue
        )
    }
}

File: lib/internal/native_models/enriched_native_error.dart
/// Enriched error information from native operations
/// Preserves structured error data for proper error handling in Dart
class EnrichedNativeError {
  const EnrichedNativeError({
    required this.message,
    required this.code,
    this.nativeError,
    this.nativeErrorCode,
    this.nativeErrorDomain,
    this.context,
    required this.timestamp,
    this.nativeStackTrace,
    required this.operationId,
    required this.instanceId,
    required this.queue,
  });

  final String message;
  final String code;
  final String? nativeError;
  final int? nativeErrorCode;
  final String? nativeErrorDomain;
  final Map<String, dynamic>? context;
  final String timestamp;
  final String? nativeStackTrace;
  final String operationId;
  final String instanceId;
  final String queue;

  /// Convert to map for sending to native
  Map<String, dynamic> toNative() {
    final map = <String, dynamic>{
      'message': message,
      'code': code,
      'timestamp': timestamp,
      'operationId': operationId,
      'instanceId': instanceId,
      'queue': queue,
    };

    if (nativeError != null) {
      map['nativeError'] = nativeError;
    }

    if (nativeErrorCode != null) {
      map['nativeErrorCode'] = nativeErrorCode;
    }

    if (nativeErrorDomain != null) {
      map['nativeErrorDomain'] = nativeErrorDomain;
    }

    if (context != null) {
      map['context'] = context;
    }

    if (nativeStackTrace != null) {
      map['nativeStackTrace'] = nativeStackTrace;
    }

    return map;
  }

  /// Convert to legacy string format for backward compatibility
  String toLegacyString() {
    final parts = <String>[message];

    if (code != 'UNKNOWN_ERROR') {
      parts.add('Code: $code');
    }

    if (nativeError != null) {
      parts.add('Native: $nativeError');
    }

    if (nativeErrorCode != null) {
      parts.add('Native Code: $nativeErrorCode');
    }

    if (nativeErrorDomain != null) {
      parts.add('Native Domain: $nativeErrorDomain');
    }

    if (context != null && context!.isNotEmpty) {
      final contextStr =
          context!.entries.map((e) => '${e.key}: ${e.value}').join(', ');
      parts.add('Context: {$contextStr}');
    }

    if (timestamp.isNotEmpty) {
      parts.add('Time: $timestamp');
    }

    if (nativeStackTrace != null) {
      parts.add('Native Stack: $nativeStackTrace');
    }

    return parts.join(' | ');
  }

  @override
  String toString() {
    return 'EnrichedNativeError(message: $message, code: $code, operationId: $operationId)';
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    return other is EnrichedNativeError &&
        other.message == message &&
        other.code == code &&
        other.operationId == operationId;
  }

  @override
  int get hashCode {
    return Object.hash(message, code, operationId);
  }
}
